### PR TITLE
Add Kleisli#mapK equivalent returning FunctionK

### PR DESCRIPTION
**Pull request by kubukoz**
Related: https://github.com/typelevel/cats-effect/pull/334

I don't like the name either, so feel free to suggest a better one 😅

### CODE REVIEW

**Comment by kubukoz** [comment]
Sure, I wanted to get some initial feedback before taking the time to write (a) test(s) :) as mentioned in the issue, I'd gladly rename it too, but `mapK` is taken (even though it's somewhere else, it clearly has different types)

**Comment by kailuowang** [comment]
This `mapK` type of operation keeps coming back, maybe we should reconsider this PR #1815. The `liftNT` (or we could name it `liftNaturalTransformation` ) provides this `mapKK` feature for all those data types through type class.

**Comment by kubukoz** [comment]
`cats-tagless`'s FunctorK would be a generalization of `mapK`, but it has
`def mapK[F[_], G[_]](af: A[F])(fk: F ~> G): A[G]`
and doesn't provide anything of the shape `A[F] ~> A[G]`.

I would start by defining this for Kleisli for now, and if we see this used by people enough to make an abstraction, maybe then we'll add a typeclass? (`FunctorK`)

**Comment by kailuowang** [comment]
I am ok with adding it to Kleisli first under the name liftK or liftNT

**Comment by kubukoz** [comment]
There's already a `liftK`. `liftNT` is free, but I'm afraid the `lift` prefix doesn't match what that function would do. After all, this operates on an already existing Kleisli, and the `lift`ing functions create Kleislis out of values like `F[B]` or `B`. Here's a potentially stupid idea: `fmapK`, just to disambiguate it from `mapK`?

**Comment by kailuowang** [comment]
The `lift` method in `Functor` is the inspiration of this name. 
```Scala
def lift[A, B](f: A => B): F[A] => F[B] =
    fa => map(fa)(f) 
```
I think `liftNT` or `liftFunctionK` is fine since it's about lifting a FunctionK to another FunctionK to the Kleisli space. `lift` is an general action, to be honest, I don't see much confusion if you have `lift` that  lifts a `F[B]` or `B` to a `Kleisli`, and `liftNT` that lifts a `FunctionK` of `F`, `G` to another FunctionK of `Kleisli[F]` and `Kleisli[G]`

**Comment by kubukoz** [comment]
That makes sense, yeah. I think `liftNT` will be good enough then :)

**Comment by kubukoz** [comment]
Renamed to `liftNT`, added a test for consistency with `mapK` and some scaladoc - let's proceed :)

**Comment by kubukoz** [comment]
Apparently the addition isn't binary compatible, but only on 2.11. Any ideas, @kailuowang?

**Comment by LukaJCB** [comment]
Make a new trait `KleisliFunctionsBinCompat`, add the function there and then mix it into `KleisliFunctions` :)

**Comment by kubukoz** [comment]
hmm, now I see that you mentioned `object Kleisli`, won't it work if I mix it into `KleisliFunctions`?

**Comment by kubukoz** [comment]
@LukaJCB hi, is there anything else I need to do with this one?

File: core/src/main/scala/cats/data/Kleisli.scala [hunk 1]
```diff
@@ -111,21 +110,21 @@ object Kleisli extends KleisliInstances with KleisliFunctions with KleisliExplic
     }
 }
 
-sealed private[data] trait KleisliFunctions {
+private[data] sealed trait KleisliFunctions {
```
Comment (LukaJCB) [inline core/src/main/scala/cats/data/Kleisli.scala]: Was this done for a particular reason? :)

File: core/src/main/scala/cats/data/Kleisli.scala [hunk 2]
```diff
@@ -111,21 +110,21 @@ object Kleisli extends KleisliInstances with KleisliFunctions with KleisliExplic
...
-   */
+    * Same as [[liftF]], but expressed as a FunctionK for use with mapK
+    * {{{
+    * scala> import cats._, data._, implicits._
+    * scala> val a: OptionT[Eval, Int] = 1.pure[OptionT[Eval, ?]]
+    * scala> val b: OptionT[Kleisli[Eval, String, ?], Int] = a.mapK(Kleisli.liftK)
+    * scala> b.value.run("").value
+    * res0: Option[Int] = Some(1)
+    * }}}
+    */
```
Comment (LukaJCB) [inline core/src/main/scala/cats/data/Kleisli.scala]: This is IntelliJ autoformating, I assume?

Comment (kubukoz) [reply core/src/main/scala/cats/data/Kleisli.scala]: ugh. For some reason it didn't enable scalafmt in cats, it seems...

**Comment by LukaJCB** [comment]
Sorry @kubukoz, it looks good, just a couple of nitpicks :)

File: core/src/main/scala/cats/data/Kleisli.scala (hunk 1 above)
Comment (kubukoz) [reply core/src/main/scala/cats/data/Kleisli.scala]: no, just a rebase issue - I'll revert

**Comment by kailuowang** [comment]
@kubukoz build failed due to scalafmt check, would you run `sbt prePR` and commit the changes?

**Comment by kubukoz** [comment]
Done :)

**Comment by kubukoz** [comment]
@kailuowang WDYT about merging this?

**Comment by kailuowang** [comment]
@kubukoz sorry I dropped the ball on this one. And after another look, I am not sure about the `liftNT` name I suggested. We renamed `NaturalTransformation` to `FunctionK` a while ago so `NT` might be confusing.  Would it be possible to rename it to `liftFK` or `liftFunctionK`? Sorry about the back and forth.

**Comment by kubukoz** [comment]
It's okay, I know how important it is to get the API right when you can't rename it for a while :)

We can definitely rename, and although I like shorter names I'll say `liftFK`'s `FK` part reminds me too much of "foreign key" and I'd go with `liftFunctionK` to be more explicit.

**Comment by LukaJCB** [comment]
I think you renamed the wrong function :P

**Comment by kubukoz** [comment]
Yeah, just noticed. Time does things to people...

**Comment by kubukoz** [comment]
(Rerunning due to the neverending Travis build on 2.13)

**Pull request merged**